Composer/packages/client/src/components/ImportModal/ImportModal.tsx

```diff
@@ -0,0 +1,338 @@
+            setImportedProjectInfo(projectInfo);
+
+            if (alias) {
+              // check to see if Composer currently has a bot project corresponding to the alias
+              res = await fetch(`/api/projects/alias/${alias}`, { method: 'GET' });
+              if (res.status === 200) {
+                const project = await res.json();
+                setExistingProject(project);
+                // ask user if they want to save to existing, or save as a new project
+                setModalState('promptingToSave');
+                return;
+              }
+            }
+            importAsNewProject(projectInfo);
+          } catch (e) {
+            // something went wrong, abort and navigate to the home page
+            console.error(`Aborting import: ${e}`);
+            navigate('/home');
+          }
+        }
+      };
+      importBotContent();
+    }
+  }, [modalState, importPayload, importSource]);
+
+  useEffect(() => {
+    if (modalState === 'signingIn') {
+      const signIn = async () => {
+        try {
+          const res = await fetch(
+            `/api/import/${importSource}/authenticate?payload=${encodeURIComponent(JSON.stringify(importPayload))}`,
+            {
+              method: 'POST',
+            }
+          );
+          if (res.status !== 200) {
+            throw `Something went wrong during authenticating import: ${res.status} ${res.statusText}`;
+          }
+          setModalState('downloadingContent');
+        } catch (e) {
+          // something went wrong, abort and navigate to the home page
+          console.error(`Aborting import: ${e}`);
+          navigate('/home');
+        }
+      };
+      signIn();
+    }
+  }, [modalState, importSource, importPayload]);
+
+  useEffect(() => {
+    if (location && location.href) {
+      try {
+        // parse data from url and store in state
+        const url = new URL(location.href);
+        const source = url.searchParams.get('source');
+        const payload = url.searchParams.get('payload');
+        if (!source || !payload) {
+          throw 'Missing source or payload.';
+        }
+        setImportSource(source);
+        setImportPayload(JSON.parse(payload));
+        setTimeout(() => {
+          setModalState('signingIn');
+        }, CONNECTING_STATUS_DISPLAY_TIME);
+      } catch (e) {
+        console.error('Aborting import: ', e);
+        navigate('/home');
+      }
+    }
+  }, []);
+
+  const cancel = useCallback(() => {
+    console.log('Cancelled import.');
+    navigate('/home');
+  }, []);
+
+  const openExistingProject = useCallback(() => {
+    navigate(`/bot/${existingProject?.id}`);
+  }, [existingProject]);
+
+  const createNewProxy = useCallback(() => {
+    if (importedProjectInfo) {
+      importAsNewProject(importedProjectInfo);
+    }
+  }, [importedProjectInfo]);
+
+  const modalContent = useMemo(() => {
+    switch (modalState) {
```

Comment:
I highly recommend to extract the larger Dialog components below to their own components, it makes code more readable, debugging in dev tools much easier and allows for better testing of each individual (dialog) component.